Index: libfprint-1.94.3+tod1/data/autosuspend.hwdb
===================================================================
--- libfprint-1.94.3+tod1.orig/data/autosuspend.hwdb
+++ libfprint-1.94.3+tod1/data/autosuspend.hwdb
@@ -148,6 +148,12 @@ usb:v04F3p0C88*
  ID_AUTOSUSPEND=1
  ID_PERSIST=0
 
+# Supported by libfprint driver elanmoc2
+usb:v04F3p0C4C*
+usb:v04F3p0C00*
+ ID_AUTOSUSPEND=1
+ ID_PERSIST=0
+
 # Supported by libfprint driver etes603
 usb:v1C7Ap0603*
  ID_AUTOSUSPEND=1
@@ -258,9 +264,7 @@ usb:v138Ap0091*
 
 # Known unsupported devices
 usb:v04F3p036B*
-usb:v04F3p0C00*
 usb:v04F3p0C4B*
-usb:v04F3p0C4C*
 usb:v04F3p0C57*
 usb:v04F3p0C5E*
 usb:v04F3p2706*
Index: libfprint-1.94.3+tod1/libfprint/fprint-list-udev-hwdb.c
===================================================================
--- libfprint-1.94.3+tod1.orig/libfprint/fprint-list-udev-hwdb.c
+++ libfprint-1.94.3+tod1/libfprint/fprint-list-udev-hwdb.c
@@ -30,9 +30,7 @@ static const FpIdEntry whitelist_id_tabl
    *   gio cat https://gitlab.freedesktop.org/libfprint/wiki/-/wikis/Unsupported-Devices.md | sed -n 's!|.*\([0-9a-fA-F]\{4\}\):\([0-9a-fA-F]\{4\}\).*|.*!  { .vid = 0x\1, .pid = 0x\2 },!p'
    */
   { .vid = 0x04f3, .pid = 0x036b },
-  { .vid = 0x04f3, .pid = 0x0c00 },
   { .vid = 0x04f3, .pid = 0x0c4b },
-  { .vid = 0x04f3, .pid = 0x0c4c },
   { .vid = 0x04f3, .pid = 0x0c57 },
   { .vid = 0x04f3, .pid = 0x0c5e },
   { .vid = 0x04f3, .pid = 0x2706 },
