From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 21 Jan 2021 15:34:36 +0100
Subject: build: Skip the hwdb test if not all drivers are really enabled

Origin: https://gitlab.freedesktop.org/libfprint/libfprint/-/commit/3b83157e9
---
 meson.build | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/meson.build b/meson.build
index 4c73c00..d5081c9 100644
--- a/meson.build
+++ b/meson.build
@@ -181,6 +181,10 @@ foreach driver: drivers
     endif
 endforeach
 
+if default_drivers_are_enabled and supported_drivers != drivers
+    default_drivers_are_enabled = false
+endif
+
 # Export the drivers' types to the core code
 drivers_type_list = []
 drivers_type_func = []
