From: hermanlin <herman.lin@emc.com.tw>
Date: Tue, 10 Aug 2021 16:34:37 +0800
Subject: elanmoc: Return the correct error when the storage is full

---
 libfprint/drivers/elanmoc/elanmoc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libfprint/drivers/elanmoc/elanmoc.c b/libfprint/drivers/elanmoc/elanmoc.c
index 2a38fbf..846d32f 100644
--- a/libfprint/drivers/elanmoc/elanmoc.c
+++ b/libfprint/drivers/elanmoc/elanmoc.c
@@ -332,7 +332,8 @@ elanmoc_reenroll_cb (FpiDeviceElanmoc *self,
       if ((self->curr_enrolled == (ELAN_MAX_ENROLL_NUM + 1)) && (buffer_in[1] == 0x00))
         {
           fp_warn ("elanmoc_reenroll_cb over enroll max");
-          fpi_ssm_mark_completed (self->task_ssm);
+          fpi_ssm_mark_failed (self->task_ssm,
+                               fpi_device_error_new (FP_DEVICE_ERROR_DATA_FULL));
           return;
         }
       if (buffer_in[1] == 0x00)
