From: ArronYen <arron.yen@emc.com.tw>
Date: Mon, 20 Jun 2022 14:03:05 +0800
Subject: elanmoc: add PID 0x0c88

---
 libfprint/drivers/elanmoc/elanmoc.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/libfprint/drivers/elanmoc/elanmoc.c b/libfprint/drivers/elanmoc/elanmoc.c
index d0dd6e8..2a38fbf 100644
--- a/libfprint/drivers/elanmoc/elanmoc.c
+++ b/libfprint/drivers/elanmoc/elanmoc.c
@@ -26,6 +26,7 @@ G_DEFINE_TYPE (FpiDeviceElanmoc, fpi_device_elanmoc, FP_TYPE_DEVICE)
 
 static const FpIdEntry id_table[] = {
   { .vid = 0x04f3,  .pid = 0x0c7e,  },
+  { .vid = 0x04f3,  .pid = 0x0c88,  },
   { .vid = 0,  .pid = 0,  .driver_data = 0 },   /* terminating entry */
 };
 
