From: hermanlin <herman.lin@emc.com.tw>
Date: Tue, 10 Aug 2021 16:45:16 +0800
Subject: elanmoc: Increase of the timeout

---
 libfprint/drivers/elanmoc/elanmoc.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libfprint/drivers/elanmoc/elanmoc.h b/libfprint/drivers/elanmoc/elanmoc.h
index daf630e..312a4b4 100644
--- a/libfprint/drivers/elanmoc/elanmoc.h
+++ b/libfprint/drivers/elanmoc/elanmoc.h
@@ -34,7 +34,7 @@ G_DECLARE_FINAL_TYPE (FpiDeviceElanmoc, fpi_device_elanmoc, FPI, DEVICE_ELANMOC,
 #define ELAN_EP_MOC_CMD_IN (0x4 | LIBUSB_ENDPOINT_IN)
 #define ELAN_EP_IMG_IN (0x2 | LIBUSB_ENDPOINT_IN)
 
-#define ELAN_MOC_CMD_TIMEOUT 2000
+#define ELAN_MOC_CMD_TIMEOUT 5000
 #define ELAN_MOC_CAL_RETRY 500
 #define ELAN_MOC_ENROLL_TIMES 9
 #define ELAN_MAX_USER_ID_LEN 92
