From: boger <boger@goodix.com>
Date: Fri, 17 Sep 2021 19:28:51 +0800
Subject: goodixmoc: add PID 63BC 63BC: Dell XPS series fingerprint sensor

Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/libfprint/+bug/1945298
---
 libfprint/drivers/goodixmoc/goodix.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libfprint/drivers/goodixmoc/goodix.c b/libfprint/drivers/goodixmoc/goodix.c
index 2197d7d..b67096e 100644
--- a/libfprint/drivers/goodixmoc/goodix.c
+++ b/libfprint/drivers/goodixmoc/goodix.c
@@ -1307,6 +1307,7 @@ gx_fp_probe (FpDevice *device)
     case 0x609C:
     case 0x639C:
     case 0x63AC:
+    case 0x63BC:
     case 0x6A94:
       self->max_enroll_stage = 12;
       break;
@@ -1526,6 +1527,7 @@ static const FpIdEntry id_table[] = {
   { .vid = 0x27c6,  .pid = 0x60A2,  },
   { .vid = 0x27c6,  .pid = 0x639C,  },
   { .vid = 0x27c6,  .pid = 0x63AC,  },
+  { .vid = 0x27c6,  .pid = 0x63BC,  },
   { .vid = 0x27c6,  .pid = 0x6496,  },
   { .vid = 0x27c6,  .pid = 0x6584,  },
   { .vid = 0x27c6,  .pid = 0x658C,  },
