From: boger <boger@goodix.com>
Date: Wed, 20 Oct 2021 20:11:05 +0800
Subject: goodixmoc: add PID 63CC 63CC: Dell Latitude series fingerprint
 sensor

---
 libfprint/drivers/goodixmoc/goodix.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libfprint/drivers/goodixmoc/goodix.c b/libfprint/drivers/goodixmoc/goodix.c
index 77394e5..f365442 100644
--- a/libfprint/drivers/goodixmoc/goodix.c
+++ b/libfprint/drivers/goodixmoc/goodix.c
@@ -1310,6 +1310,7 @@ gx_fp_probe (FpDevice *device)
     case 0x639C:
     case 0x63AC:
     case 0x63BC:
+    case 0x63CC:
     case 0x6A94:
       self->max_enroll_stage = 12;
       break;
@@ -1530,6 +1531,7 @@ static const FpIdEntry id_table[] = {
   { .vid = 0x27c6,  .pid = 0x639C,  },
   { .vid = 0x27c6,  .pid = 0x63AC,  },
   { .vid = 0x27c6,  .pid = 0x63BC,  },
+  { .vid = 0x27c6,  .pid = 0x63CC,  },
   { .vid = 0x27c6,  .pid = 0x6496,  },
   { .vid = 0x27c6,  .pid = 0x6584,  },
   { .vid = 0x27c6,  .pid = 0x658C,  },
