From: mahaosen <mahaosen@goodix.com>
Date: Fri, 26 Aug 2022 16:46:16 +0800
Subject: goodixmoc: add PID 0x634C

---
 libfprint/drivers/goodixmoc/goodix.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libfprint/drivers/goodixmoc/goodix.c b/libfprint/drivers/goodixmoc/goodix.c
index f365442..d4fa432 100644
--- a/libfprint/drivers/goodixmoc/goodix.c
+++ b/libfprint/drivers/goodixmoc/goodix.c
@@ -1307,6 +1307,7 @@ gx_fp_probe (FpDevice *device)
     case 0x6496:
     case 0x60A2:
     case 0x609C:
+    case 0x634C:
     case 0x639C:
     case 0x63AC:
     case 0x63BC:
@@ -1528,6 +1529,7 @@ static const FpIdEntry id_table[] = {
   { .vid = 0x27c6,  .pid = 0x5840,  },
   { .vid = 0x27c6,  .pid = 0x609C,  },
   { .vid = 0x27c6,  .pid = 0x60A2,  },
+  { .vid = 0x27c6,  .pid = 0x634C,  },
   { .vid = 0x27c6,  .pid = 0x639C,  },
   { .vid = 0x27c6,  .pid = 0x63AC,  },
   { .vid = 0x27c6,  .pid = 0x63BC,  },
