From: Benjamin Berg <bberg@redhat.com>
Date: Tue, 19 Jan 2021 13:29:17 +0100
Subject: udev-rules: Print warning if an ID is supported

---
 libfprint/fprint-list-udev-rules.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/libfprint/fprint-list-udev-rules.c b/libfprint/fprint-list-udev-rules.c
index ff8fd1c..08b79d9 100644
--- a/libfprint/fprint-list-udev-rules.c
+++ b/libfprint/fprint-list-udev-rules.c
@@ -111,6 +111,7 @@ static const FpIdEntry blacklist_id_table[] = {
 static const FpDeviceClass whitelist = {
   .type = FP_DEVICE_TYPE_USB,
   .id_table = whitelist_id_table,
+  .id = "whitelist",
   .full_name = "Hardcoded whitelist"
 };
 
@@ -136,13 +137,15 @@ print_driver (const FpDeviceClass *cls)
 
       key = g_strdup_printf ("%04x:%04x", entry->vid, entry->pid);
 
-      if (g_hash_table_lookup (printed, key) != NULL)
+      if (cls == &whitelist && g_hash_table_lookup (printed, key) != NULL)
         {
+          g_warning ("%s implemented by driver %s",
+                     key, (const char *) g_hash_table_lookup (printed, key));
           g_free (key);
           continue;
         }
 
-      g_hash_table_insert (printed, key, GINT_TO_POINTER (1));
+      g_hash_table_insert (printed, key, (void *) cls->id);
 
       if (num_printed == 0)
         g_print ("# %s\n", cls->full_name);
